File: src/main/java/me/testaccount666/serversystem/clickablesigns/executables/weather/ConfiguratorWeatherSign.java
package me.testaccount666.serversystem.clickablesigns.executables.weather;

import me.testaccount666.serversystem.clickablesigns.AbstractSignConfigurator;
import me.testaccount666.serversystem.clickablesigns.SignType;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.utils.ChatColor;
import me.testaccount666.serversystem.utils.ComponentColor;
import org.bukkit.block.Sign;
import org.bukkit.block.sign.Side;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import static me.testaccount666.serversystem.utils.MessageBuilder.sign;

public class ConfiguratorWeatherSign extends AbstractSignConfigurator {

    @Override
    protected String getCreatePermissionNode() {
        return "ClickableSigns.Weather.Create";
    }

    @Override
    protected SignType getSignType() {
        return SignType.WEATHER;
    }

    @Override
    protected boolean validateConfiguration(User user, Sign sign, YamlConfiguration config) {
        var front = sign.getSide(Side.FRONT);
        var weatherType = front.getLine(1).toLowerCase();
        if (weatherType.isEmpty()) {
            sign("Weather.NoWeatherSpecified", user).build();
            return false;
        }

        if (!isValidWeatherType(weatherType)) {
            sign("Weather.InvalidWeather", user)
                    .postModifier(message -> message.replace("<WEATHER>", weatherType)).build();
            return false;
        }

        front.line(0, ComponentColor.translateToComponent(SignType.WEATHER.signName()));
        front.line(1, ComponentColor.translateToComponent("&2${weatherType}"));
        var back = sign.getSide(Side.BACK);
        for (var index = 0; index < 4; index++) back.line(index, front.line(index));
        sign.update();
        return true;
    }

    @Override
    protected void addSignSpecificConfiguration(User user, Sign sign, FileConfiguration config) {
        var weatherType = sign.getSide(Side.FRONT).getLine(1).toLowerCase();
        weatherType = ChatColor.stripColor(weatherType);
        config.set("WeatherType", weatherType);
    }

    @Override
    protected String getSuccessMessageKey() {
        return "Weather.Created";
    }

    private boolean isValidWeatherType(String weatherType) {
        return weatherType.equals("sun") ||
                weatherType.equals("clear") ||
                weatherType.equals("storm") ||
                weatherType.equals("thunder") ||
                weatherType.equals("rain");
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/clickablesigns/executables/weather/ConfiguratorWeatherSign.kt
package me.testaccount666.serversystem.clickablesigns.executables.weather

import me.testaccount666.paperktx.colors.ChatColor.Companion.stripColor
import me.testaccount666.paperktx.extensions.ChecksExtensions.isAny
import me.testaccount666.paperktx.extensions.ComponentExtensions.asComponent
import me.testaccount666.serversystem.clickablesigns.AbstractSignConfigurator
import me.testaccount666.serversystem.clickablesigns.SignType
import me.testaccount666.serversystem.extensions.signMsg
import me.testaccount666.serversystem.userdata.User
import org.bukkit.block.Sign
import org.bukkit.block.sign.Side
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration

class ConfiguratorWeatherSign : AbstractSignConfigurator() {
    override val createPermissionNode = "ClickableSigns.Weather.Create"
    override val successMessageKey = "Weather.Created"
    override val signType = SignType.WEATHER

    override fun validateConfiguration(user: User, sign: Sign, config: YamlConfiguration): Boolean {
        val front = sign.getSide(Side.FRONT)
        val weatherType = front.getLine(1).lowercase()
        if (weatherType.isEmpty()) {
            user.signMsg("Weather.NoWeatherSpecified")
            return false
        }

        if (!isValidWeatherType(weatherType)) {
            user.signMsg("Weather.InvalidWeather") {
                postModifier { it.replace("<WEATHER>", weatherType) }
            }
            return false
        }

        front.line(0, SignType.WEATHER.signName.asComponent())
        front.line(1, "&2${weatherType}".asComponent())
        val back = sign.getSide(Side.BACK)
        for (index in 0..3) back.line(index, front.line(index))
        sign.update()
        return true
    }

    override fun addSignSpecificConfiguration(user: User, sign: Sign, config: FileConfiguration) {
        var weatherType = sign.getSide(Side.FRONT).getLine(1).lowercase()
        weatherType = stripColor(weatherType)
        config["WeatherType"] = weatherType
    }

    private fun isValidWeatherType(weatherType: String) = weatherType.isAny("sun", "clear", "storm", "thunder", "rain")
}